### ISSUE TITLE
[Feature] Endpoint for parallel connection and authentication only

### ISSUE THREAD

**Issue by algsupport** [ParallelSSH/parallel-ssh#145]
Hello,
Is there a way to know if the connection is established or not? In some cases I have connections with very high delay and I would like to know that the connection is established before running a command.
I hope this makes sense.
In fabric I can run connection.open() and then connection.isconnected and I will get a bool that will indicate if the connection is open or not.
Is there something similar in the native client?
Thank you.

**Comment by pkittenis** [comment]
Hi there,

Thanks for the interest.

Meaning connecting then executing separately? Currently not with a public API. Both connection and command execution are handled by `run_command` which raises exceptions on any errors connecting. 

The logger will also log when connections are established before running command - see `pssh.utils.enable_logger`.

Checking connection for a particular host before running commands would necessarily have to be blocking and serial by host which [conflicts with this project's goals](https://github.com/ParallelSSH/parallel-ssh#design-and-goals). Fabric does not initiate connections in parallel AFAIK.

It can be done by using the internal `ParallelSSHClient._make_ssh_client(host)` where `host` is the host to connect to, but will block other commands from running in parallel unless gevent spawning and handling is also done client side.

Eg:

```
connection_let = gevent.spawn(client._make_ssh_client, host)
# Will raise exceptions on errors connecting/authenticating
connection_let.get()
```

A new public API function could be made to establish connections in parallel for all hosts separately from running command, for such use cases. Will leave this open to track that.

**Comment by teucrus** [comment]
I mostly want to cover the following scenario:
I'd like to be able to keep the connection open and send multiple commands on the same connection/session. This is mostly because the servers I manage might have a delay of up to 4 seconds and I don't want to make a new connection for each command.

**Comment by pkittenis** [comment]
That use case is already handled by `run_command`. 

Multiple commands for a previously used host on the same client use the same established connection and authenticated session. [Per documentation](https://parallel-ssh.readthedocs.io/en/latest/quickstart.html#run-a-command-on-hosts-in-parallel):

> The client object can, and should, be reused. Existing connections to hosts will remain alive as long as the client object is kept alive. Subsequent commands to the same host(s) will reuse their existing connection and benefit from much faster response times.

**Comment by teucrus** [comment]
> That use case is already handled by `run_command`.
> 
> Multiple commands for a previously used host on the same client use the same established connection and authenticated session. [Per documentation](https://parallel-ssh.readthedocs.io/en/latest/quickstart.html#run-a-command-on-hosts-in-parallel):
> 
> > The client object can, and should, be reused. Existing connections to hosts will remain alive as long as the client object is kept alive. Subsequent commands to the same host(s) will reuse their existing connection and benefit from much faster response times.

And what happens if I need to close the connection at some part of my script?

**Comment by pkittenis** [comment]
Per documentation:

> Existing connections to hosts will remain alive as long as the client object is kept alive.

If client connections are not needed anymore, `del client` or let it go out of scope.

**Comment by pkittenis** [comment]
`2.2.0` has `connect_auth` to do this.
